rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,511 @@
+      newer documents are present in the index and refuse to start up.
+    2. Set the outdated index to read-only. Since `.kibana` is never advanced,
+      it will be pointing to a read-only index which prevent writes from
+      6.8+ releases which are already online.
+</details>
+
+## 4.5 Changes in 8.0
+1. Migrations will be in-place and re-use the same index for minor upgrades.
+2. Only allow breaking field mapping changes (which requires creating a new
+   index) in a major.
+3. All update operations should use optimistic concurrency control.
+
+### 4.5.1 Migration algorithm (8.0):
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed.
+2. If there are documents with newer `migrationVersion` numbers, exit Kibana
+   with a fatal error.
+3. If there are outdated documents, migrate these in batches:
+   1. Read a batch of outdated documents from `.kibana_n`.
+   2. Transform documents by applying the migration functions.
+   3. Update the document batch in the same index using optimistic concurrency
+      control. 
+   4. If a batch fails due to an update version mismatch, repeat the entire
+      process.
+4. Once all documents are up to date, the migration is complete and Kibana can
+   start serving traffic.
+
+Advantages:
+- Not duplicating all documents into a new index will speed up migrations and
+  reduce the downtime window. This will be especially important for the future
+  requirement to support > 10k or > 100k documents.
+- We can check the health of an existing index before starting the migration,
+  but we cannot detect what kind of failures might occur while creating a new
+  index. Whereas retrying migrations will eventually recover from the errors
+  in (3.3), re-using an index allows us to detect these problems before trying
+  and avoid errors like (3.3.1) altogether.
+- Single index to backup instead of “index pattern” that matches any
+  `.kibana_n`.
+- Simplifies Kibana system index Elasticsearch plugin since it needs to work
+  on one index per "tenant".
+- Can prevent data loss when mixed versions of Kibana are both attempting a
+  migration using optimistic concurrency control.
+
+Drawbacks:
+- Cannot make breaking mapping changes (even though it was possible we never
+  introduced a breaking mapping change during 7.x).
+- Rollback is only possible by restoring a snapshot which requires educating
+  users to ensure that they don't rely on `.kibana_n` indices as backups.
+  (Apart from the need to educate users, snapshot restores provide many
+  benefits).
+- It narrows the second restriction under (4.3.1) even further: migrations
+  cannot rely on any state that could change as part of a migration because we
+  can no longer use the previous index as a snapshot of unmigrated state.
+- We can’t automatically recover from a half-way done migration.
```

Comment:
Not sure I understand this limitation. Why would a migration that failed half way not be able to be completed? Wouldn't this limitation eliminate the possibility of automated retries? (the primary source of support cases)

Reply:
The wording here is misleading (I'll fix it), the drawback is that we can't automatically rollback from a failed migration. So once you start you either need a snapshot or find a way to get the migration to complete successfully.